myia/compile/backends/python/optimizer.py

```diff
@@ -0,0 +1,105 @@
+"""Python backend optimized.
+
+Optimize directed graphs just before code generation.
+
+Inputs: sequence of directed graphs to compile.
+
+Output: a list of special dictionaries to pass to code generator to "patch" directed graphs.
+Neither directed graphs nor myia graphs will be modified. Instead, code generator will use
+"patch" dictionaries to generate optimized code on-the-fly.
+"""
+from myia.compile.backends.python.directed_graph import DirectedGraph
+from myia.ir.node import Apply
+
+TYPEOF = "typeof"
+MAKE_HANDLE = "make_handle"
+UNIVERSE_SETITEM = "universe_setitem"
+UNIVERSE_GETITEM = "universe_getitem"
+
+
+class Optimizer:
+    """Optimizer class."""
+
+    def __init__(self):
+        self.skip = {}
+        self.replace = {}
+        self.rename = {}
+        self.nonlocals = {}
+
+    def optimize(self, directed_graphs):
+        """Optimize directed graphs.
+
+        :param directed_graphs: list of directed graphs to compile.
+        :return: patch: a map of dictionaries to pass to code generator to apply optimization during code generator.
+        """
+        # Recursively optimize directed graphs.
+        todo = list(directed_graphs)
+        seen = set()
+        while todo:
+            dg = todo.pop(0)
+            # Each directed graph should be visited once.
+            assert dg not in seen
+            seen.add(dg)
+            self._optimize_universe_setitem(dg)
+            self._optimize_universe_getitem(dg)
+            todo.extend(
+                element
+                for element in dg.value_to_node
+                if isinstance(element, DirectedGraph)
+            )
+        return {
+            "skip": self.skip,
+            "replace": self.replace,
+            "rename": self.rename,
+            "nonlocals": self.nonlocals,
+        }
+
+    def _optimize_universe_setitem(self, dg: DirectedGraph):
+        """Optimize `universe_setitem` nodes for given directed graph."""
+        # Dictionary mapping typename to myia node to assignation object
+        nodes = {TYPEOF: set(), MAKE_HANDLE: set(), UNIVERSE_SETITEM: set()}
+        for element in dg.value_to_node:
+            if isinstance(element, Apply):
+                fn = element.fn
+                if fn.is_constant() and fn.value in nodes:
+                    assert element not in nodes[fn.value]
+                    nodes[fn.value].add(element)
```

Comment:
I think the design for this could be made somewhat more generic, by e.g. having a function like `self.collect_nodes({TYPEOF, MAKE_HANDLE, UNIVERSE_SETITEM})` return the nodes dict you have here. That function could be reused for `_optimize_universe_getitem`, with a different argument.

Reply:
Ok !